network/src/commonJvmAndroid/kotlin/com/wire/kalium/network/HttpEngine.kt

```diff
@@ -3,33 +3,42 @@ package com.wire.kalium.network
 import io.ktor.client.engine.HttpClientEngine
 import io.ktor.client.engine.okhttp.OkHttp
 import okhttp3.Interceptor
+import okhttp3.Request
 import okhttp3.Response
 
 actual fun defaultHttpEngine(): HttpClientEngine {
     return OkHttp.create {
-        /**
-         * This is a hack
-         *
-         * As seen here: https://github.com/ktorio/ktor/issues/1127
-         * Ktor is pretty has `application/protobuf` but not `application/x-protobuf`
-         * and it doesn't support custom content types!
-         */
         addInterceptor(object : Interceptor {
             override fun intercept(chain: Interceptor.Chain): Response {
-                val currentRequest = chain.call().request()
-                val currentHeaders = chain.call().request().headers
-                // When sending a byteArray, Ktor will put `octet-stream`. We can identify this and replace with protobuf
-                val isProtoBuf = currentHeaders["Content-Type"] == "application/octet-stream"
-
-                return if (!isProtoBuf) {
-                    chain.proceed(currentRequest)
-                } else {
-                    val newRequest = currentRequest.newBuilder()
-                        .header("Content-Type", "application/x-protobuf")
-                        .build()
-                    chain.proceed(newRequest)
-                }
+                return chain.call().request().handleProtoHeader()
+                    .let { checkedRequest -> chain.proceed(checkedRequest) }.handleUnauthorizedResponse()
             }
         })
     }
 }
+
+/**
+ * This is a hack
+ *
+ * As seen here: https://github.com/ktorio/ktor/issues/1127
+ * Ktor is pretty has `application/protobuf` but not `application/x-protobuf`
+ * and it doesn't support custom content types!
+ */
+private fun Request.handleProtoHeader(): Request =
+    when (headers["Content-Type"] == "application/octet-stream") {
+        true -> newBuilder().header("Content-Type", "application/x-protobuf").build()
```

Comment:
I might not be reading it correctly, but wouldn't this overwrite any other headers that the original might contain??

Reply:
Yes, this will overwrite the Content-Type header in case it's octet-stream
the issue is when sending encrypted blob of data ktor will set the Content-Type to octet-stream, but BE needs x-protobuf.

Reply:
I get that, but what about other headers that the original request has? Wouldn't `newBuilder().header("Content-Type", "application/x-protobuf").build()` just create a new request with just one content type header??

Reply:
Maybe the style is a bit confusing since it's an extension function, but the `newBuilder().header(......` is equivalent to `currentRequest.newBuilder().header(.......` 
OkHttp Docs:
>   * Sets the header named [name] to [value]. If this request already has any headers
     * with that name, they are all replaced.
    open fun header(name: String, value: String) = apply {
      headers[name] = value
    }

so no headers lost in the process